universal-hash/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+#![warn(missing_docs, rust_2018_idioms)]
+
+#[cfg(feature = "std")]
+extern crate std;
+
+use generic_array::{ArrayLength, GenericArray};
+use generic_array::typenum::Unsigned;
+use subtle::{Choice, ConstantTimeEq};
+
+/// The `UniversalHash` trait defines a generic interface for universal hash
+/// functions.
+pub trait UniversalHash: Clone {
+    /// Size of a block (e.g. field element) this universal hash operates on
+    type BlockSize: ArrayLength<u8>;
+
+    /// Instantiate a universal hash function with the given key
+    fn new(key: &GenericArray<u8, Self::BlockSize>) -> Self;
+
+    /// Input a block into the universal hash function
+    fn update_block(&mut self, block: &GenericArray<u8, Self::BlockSize>);
+
+    /// Input data into the universal hash function. If the length of the
+    /// data is not a multiple of the block size, the remaining data is
+    /// padded with zeroes up to the `BlockSize`.
+    ///
+    /// This approach is frequently used by AEAD modes which use
+    /// Message Authentication Codes (MACs) based on universal hashing.
+    fn update_padded(&mut self, data: &[u8]) {
+        let mut chunks = data.chunks_exact(Self::BlockSize::to_usize());
+
+        for chunk in &mut chunks {
+            self.update_block(GenericArray::from_slice(chunk));
+        }
+
+        let rem = chunks.remainder();
+
+        if !rem.is_empty() {
+            let mut padded_block = GenericArray::default();
+            padded_block[..rem.len()].copy_from_slice(rem);
+            self.update_block(&padded_block);
+        }
+    }
+
+    /// Reset `UniversalHash` instance.
+    fn reset(&mut self);
+
+    /// Obtain the [`Output`] of a `UniversalHash` function and consume it.
+    fn result(self) -> Output<Self::BlockSize>;
+
+    /// Obtain the [`Output`] of a `UniversalHash` computation and reset it back
+    /// to its initial state.
+    fn result_reset(&mut self) -> Output<Self::BlockSize> {
+        let res = self.clone().result();
+        self.reset();
+        res
+    }
+
+    /// Verify the `UniversalHash` of the processed input matches a given [`Output`].
+    /// This is useful when constructing Message Authentication Codes (MACs)
+    /// from universal hash functions.
+    fn verify(self, other: impl Into<Output<Self::BlockSize>>) -> Result<(), Error> {
+        if self.result() == other.into() {
+            Ok(())
+        } else {
+            Err(Error)
+        }
+    }
+}
+
+/// Outputs of universal hash functions which are a thin wrapper around a
+/// byte array. Provides a safe `Eq` implementation that runs in constant time,
+/// which is useful for implementing Message Authentication Codes (MACs) based
+/// on universal hashing.
+#[derive(Clone)]
+pub struct Output<N: ArrayLength<u8>> {
+    bytes: GenericArray<u8, N>,
+}
+
+impl<N> Output<N>
+where
+    N: ArrayLength<u8>,
+{
+    /// Create a new `Block`.
+    pub fn new(bytes: GenericArray<u8, N>) -> Output<N> {
+        Output { bytes }
+    }
```

Comment:
Note that we don't have any [`code`](https://docs.rs/crypto-mac/0.7.0/crypto_mac/struct.MacResult.html#method.code) method to retrieve result bytes. If I am not missing something it would make this type unusable if we want to transfer or store the resulting tag.

Reply:
Good catch!

I'm not sure `code` is the appropriate name here, but I do agree there needs to be a way to do this.

How about `into_bytes` and/or an appropriate `Into` impl?

Reply:
I think a dedicated method will be easier to discover. `into_bytes` sounds good (we probably will need to rename `MacResult` method for consistency later). `Into` can be added as well if you think it will be useful for users.

Reply:
Ok, `Output::into_bytes` added in 83b438f